src/charmcraft/charmcraft.js

```diff
@@ -0,0 +1,109 @@
+const core = require("@actions/core");
+const exec = require("@actions/exec");
+const fs = require("fs");
+const yaml = require("js-yaml");
+const glob = require("@actions/glob");
+
+/* eslint-disable camelcase */
+
+class Charmcraft {
+  constructor() {
+    this.uploadImage = core.getInput("upload-image").toLowerCase() === "true";
+    core.exportVariable("CHARMCRAFT_AUTH", core.getInput("credentials"));
+  }
+
+  async uploadResources() {
+    let resourceInfo = "resources:\n";
+
+    const { name, images } = this.metadata();
+    const flags = await Promise.all(
+      images.map(async ([resource_name, resource_image]) => {
+        await this.uploadResource(resource_image, name, resource_name);
+        const { flag, info } = await this.buildResourceFlag(
+          name,
+          resource_name,
+          resource_image
+        );
+        resourceInfo += info;
+        return flag;
+      })
+    );
+    return { flags, resourceInfo };
+  }
+
+  async uploadResource(resource_image, name, resource_name) {
+    if (!this.uploadImage) {
+      core.warning(
+        "No resources where uploaded as part of this build.\n" +
+          `If you wish to upload the OCI image, set 'upload-image' to 'true'`
+      );
+      return;
+    }
+
+    await exec.exec("docker", ["pull", resource_image]);
```

Comment:
This does make the assumption of access to the Docker socket through group membership or similar - is it worth catching the case where that doesn't happen? I kinda wish we could fall back to something lighter like `skopeo` in these cases

Reply:
Makes sense! If the charmhub registry supports head requests that would be even better and not require us to shuffle all that data around just to immediately trash it afterwards. I will need to have a look however.

Reply:
Spit into issue #22